sh-runtime.c

```diff
@@ -673,115 +544,97 @@ DEPENDS_ON(putchar)
   putstr("    : $((__addr += 1))\n");
   putstr("  done\n");
   putstr("}\n");
-END_RUNTIME_FUN(print_pnut_str)
+END_RUNTIME_FUN(put_pstr)
 
-DEFINE_RUNTIME_FUN(print_string)
-DEPENDS_ON(int_to_char)
-  putstr("# Emit a C-string line by line so that whitespace isn't mangled\n");
-  putstr("print_string() {\n");
-  putstr("  __addr=$1; shift\n");
-  putstr("  __max_len=100000000\n");
-  putstr("  __delim=0\n");
-  putstr("  __len=0\n");
-  putstr("  __acc=\"\"\n");
-  putstr("  if [ $# -ge 1 ] ; then __delim=$1   ; shift ; fi # Optional end of string delimiter\n");
-  putstr("  if [ $# -ge 1 ] ; then __max_len=$1 ; shift ; fi # Optional max length\n");
-  putstr("  while [ $((_$__addr)) != $__delim ] && [ $__max_len -gt $__len ] ; do\n");
-  putstr("    __char=$((_$__addr))\n");
-  putstr("    __addr=$((__addr + 1))\n");
-  putstr("    __len=$((__len + 1))\n");
-  putstr("    case $__char in\n");
-  putstr("      10) # 10 == '\\n'\n");
-  putstr("        printf \"%s\\n\" \"$__acc\"\n");
-  putstr("        __acc=\"\" ;;\n");
-  putstr("      *)\n");
-  call_int_to_char("        ", "$__char")
-  putstr("        __acc=\"$__acc$__char\" ;;\n");
-  putstr("    esac\n");
+DEFINE_RUNTIME_FUN(printf)
+DEPENDS_ON(put_pstr)
+  putstr("read_int() {\n");
```

Comment:
`pack_string` was called in a few places to do the equivalent of reading a number. This function is much simpler and easier to understand.